design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -0,0 +1,372 @@
+transient query is simply returning tabular data and has no concept of a key column, persistent
+queries do have the concept of a key column, and changing that column is only allowed via an
+explicit `PARTITION BY` or `GROUP BY` statement.  Therefore users will _always_ have to specify the
+key column in the projection of persistent queries.
+
+For example, the following statements, which previously executed, will now fail:
+
+```sql
+CREATE TABLE OUTPUT AS SELECT V0, V1 FROM INPUT;
+-- fails as key column not in projection.
+
+CREATE TABLE OUTPUT AS SELECT COUNT(*) FROM INPUT GROUP BY V1;
+-- fails as key column not in projection.
+
+CREATE STREAM OUTPUT AS SELECT V0, ID FROM INPUT PARTITION BY V1;
+-- fails as key column not in projection.
+
+CREATE TABLE OUTPUT AS SELECT I1.V0, I2.V2 FROM INPUT I1 JOIN INPUT I2 ON I1.ID = I2.ID;
+-- fails as key column not in projection.
+```
+
+We propose this is acceptable, and preferable to the current model.
+
+
+### The thorn in the design's side
+
+Hopefully, all of the above seems simple and clear. Now for the tricky and murky bit...
+
+The observant among you may be thinking 'ah, but what if I _want_ to put a copy of the key column
+into the value?'. With current syntax you can do:
+
+```sql
+-- current syntax:
+CREATE TABLE OUTPUT AS SELECT V0, V1, ID AS V2 FROM INPUT;
+-- resulting schema: ID INT KEY, V0 INT, V1 INT, V2 INT
+```
+
+With the proposed syntax the `ID AS V2` would be treated an aliased key column, resulting in the
+schema 'V2 INT KEY, V0 INT, V1 INT'. So, how does the user create an copy of the key in the value,
+if that's what they need to do?
+
+Outside of introducing some non-standard keyword into the statement, e.g.
+
+```sql
+-- KEY keyword to indicate its a KEY column
+CREATE TABLE OUTPUT AS SELECT ID AS V2 KEY, V0, V1 FROM INPUT;
+-- resulting schema: V2 INT KEY, V0 INT, V1 INT
+
+-- VALUE keyword to indicate its a VALUE column
+CREATE TABLE OUTPUT AS SELECT ID, V0, V1, ID AS V2 VALUE FROM INPUT;
+-- resulting schema: ID INT KEY, V0 INT, V1 INT, V2 INT
+```
+
+Which are pretty yuck as its moving away from standard SQL.
+
+Or by adding some `WITH(KEY_COLUMNS=?)` style syntax, e.g.
+
+```sql
+CREATE TABLE OUTPUT WITH(KEY_COLUMNS=[ID]) AS SELECT ID, V0, V1, ID AS V2 VALUE FROM INPUT;
+-- resulting schema: ID INT KEY, V0 INT, V1 INT, V2 INT
+```
+
+Which is better, but still a little verbose. Also, it would be nice to not couple this KLIP to any
+discussion about 'Should ksqlDB have `KEY` columns in the schema of CSAS statements or list them in
+the WITH clause?'
+
+Aside from these, the only ways we can think of that are standard SQL would to treat the first
+occurrence of the key column in the projection as naming the key column in the result, and any
+subsequent copies to be value columns, (see example below). However, this reliance on the _order_
+of columns is less than ideal. Anyone got any better ideas for a short term fix for this edge case?
```

Comment:
👎 relying on order is super confusing. how often are people going to be copying keys into the value? if it's OK for it to be a little clunky, we can introduce a `IDENTITY(foo)` or `COPY(foo)` UDF that just returns the same value as the input. that way you could always do:
```sql
CREATE TABLE OUTPUT AS SELECT ID, V0, V1, COPY(ID) AS V2 FROM INPUT
```
I admit that feels like a workaround, but I don't think it's awful and it fits nicely into the existing model

Reply:
I didn't like the ordering either!

The idea being that this function would mean the second `ID` in the projection is no longer a column reference, and therefore can not be confused with the key.

Not really intuitive, but probably more intuitive that ordering!

Reply:
Given that the use case does not seem to be super important, the idea of an UDF is actually quite neat IMHO. It does not seem to be worth to introduce a new non-standard keyword for this case.

Reply:
Totally with you on not wanting to rely on ordering, or new key words. Agree this will _hopefully_ be an uncommon use case, but not one can ignore.  UDF sounds good, assuming no one else has any brighter ideas! :D

Reply:
updated the KLIP.